### ISSUE TITLE
☂️ Initial internal feedback

### ISSUE THREAD

**Issue by litherum** [gpuweb/gpuweb#2213]
Here's some feedback that one of the internal teams had when starting out with WGSL:

1. **Explicit post-fix requirements**: It’s a bit surprising to have to stick post-fix modifiers on so many numbers in order to get simple operations to work: `vert.VertexIndex % 3u == 0u` https://github.com/gpuweb/gpuweb/issues/739 
2. **Lack of promotion to floating point**: It's a bit weird to also have to explicitly add `.0` to numbers to make them floating point: `vec4<f32>(1, 0, 0, 1)` must be `vec4<f32>(1.0, 0.0, 0.0, 1.0)` https://github.com/gpuweb/gpuweb/issues/739
3. **Post-fix prohibitions**: It's even weird to be prohibited from point a `f` at the back of floating given the above requirements. It's not obvious how doubles would be supported. https://github.com/gpuweb/gpuweb/issues/2210
4. **Explicit brace requirements on 1-liners**: Surprising. https://github.com/gpuweb/gpuweb/issues/2211
5. `elseif` instead of `else if`: Okay, but why? https://github.com/gpuweb/gpuweb/issues/2212

Summary: Shaders authors expect a C-like syntax, even Metal being C++ like still supports C style syntax. WGSL is surprising in that it seems to be neither fully C-style nor C++-style.

### Update

6. C-style casts don't work: `output.position.x = (float)(vert.VertexID / 2u) * 4.0 - 1.0;` https://github.com/gpuweb/gpuweb/issues/2220

**Comment by archo5** [comment]
> If you expect to translate from HLSL into WGSL using a simple preprocessor

well, it's not just me:

https://github.com/hjwdzh/Fluid3D/blob/master/fxaa.hlsl#L641
https://github.com/iryoku/smaa/blob/master/SMAA.hlsl#L528

> Instead, we are developing tools to do the translation from SPIR-V, GLSL, and potentially other languages.

what's better than having tools is not needing them.

I hope you do at least realize that there will be projects who'll need to support WebGL and WebGPU (**and their corresponding shading languages**) in parallel for many years to come
and that some of them will need to at least strongly consider using a runtime shader translator

if things that would provably make their lives easier, or provide more options for dealing with things in more effective ways are "non-goals" without some really good reason*, I'm franky not even sure what to say

*-and no, "this would limit us" is not a particularly good reason - all designers have constraints they're dealing with

it's actually scary I haven't seen much info about them (no internal performance or usability concerns), just that the language is intended to basically be SPIR-V without actually being SPIR-V 🤔 talk about "non-goal" 😄

**Comment by juj** [comment]
> If you expect to translate from HLSL into WGSL using a simple preprocessor

While that strategy may read very hacky to people who do invest their time into building and/or deploying actual converters/compilers, it is something that is very appealing, because of the whole

> what's better than having tools is not needing them.

part. Not needing to deploy (often bloated) offline (or worse, online) shader language converter toolchains/compilers, and instead utilizing an ad hoc language conventions with preprocessor to co-target multiple languages can be quite a lean path. It may not scale to large projects, but many sites are deploying moderate/constrained sets of shaders, hence being able to get away with this kind of a to-the-point solution.

I am looking at a number of WebGL codebases where I'd like to dual target WebGL and WebGPU shaders with least amount of maintenance pain and minimal increase in deployed code size and, and am pondering on what is the most sensible solution to take. WebGPU looks so alien that a preprocessor convention that works really well for co-targeting HLSL vs GLSL vs ESSL vs WebGL ESSL does not seem to work at all here.

Overall I wish that WGSL would have been built on top of C as much as possible, since it is the de facto existing knowledge base as all existing shader languages do that. Mixing in bits from JS, TS and Rust -like constructs to make it more "webby" seems like a mistake to me. It is looking like it will be alien both to existing graphics developers, and to existing web developers.

However I don't wish to put anyone's work down. I do think there would be a lot of value for the syntax to be softened to look more familiar to existing knowledge. And still hope that WGSL arrives at a good place for "lightweight cross-targeting" from some kind of C-like syntax that one can write a super-lightweight (e.g. on the order of ~50KB shipped code size) library to tokenize and on-the-fly convert to WebGL ESSL and WebGPU.

**Comment by juj** [comment]
> we are developing tools to do the translation from SPIR-V, GLSL, and potentially other languages.

One question that exists with translators is how to deal with a number of shader variations implemented in the form of a set of preprocessor variables from a single shader file. Are there translators that would retain preprocessor conditionals over to WGSL?

**Comment by magcius** [comment]
My thoughts on the whole situation:

While it is true that GLSL and HLSL are "compatible enough" in simple cases that you can mostly write them in a preprocessor style, unfortunately, MSL *really* breaks down for this use case, and tools really start to become necessary there. As much as I appreciate the language design of MSL from an independent standpoint, I do admit it's a very difficult language to work with from an ecosystem perspective, and I think that was a mistake from the MSL designers.

However, when getting into more complicated shaders, the polyglot style becomes harder to adopt. I don't think I've ever seen anybody adopt polyglot style for compute shaders; GLSL and HLSL both have very different ways of doing compute shaders, and there's no directly obvious 1:1 translation. (should HLSL's RWByteAddressBuffer map to an SSBO of bytes, or image load/store? RWBuffer with format reinterpretation only maps to image load/store with EXT_shader_image_load_formatted enabled). As engines start to develop more compute-based tools, their shader compilation and tools needs grow.

Additionally, WebGPU breaks from WebGL in a very important area: separate image/samplers. I offer no specific opinion on whether this was a good idea or a bad idea, just that it means that you can't easily compile WebGL shaders to WebGPU shaders without a substantial step already to handle this difference.

There's a question about whether WGSL should end up with a preprocessor in general, and I have conflicting feelings on that: One, I can't deny how much help a preprocessor has brought me in my own WebGL projects. But it's a *very* clunky tool not fit for very many purposes, leaving me to write my own preprocessor in a lot of cases. I'd perhaps like to see alternate tools for the use cases we need a preprocessor for, but, also, I don't believe the web is a place for new R&D for things like that, we should be solidifying ideas used in industry, and a preprocessor is what's used there.

This whole situation I feel is, uh, suboptimal, to say the least, but I don't see a clear path forward towards hitting all goals that people want.

**Comment by jimblandy** [comment]
If you're curious about how the spec arrived at its present state, there's some history here: http://kvark.github.io/spirv/2021/05/01/spirv-horrors.html  But this issue isn't the place to revisit decisions made long ago, after extensive discussion. 

Comments should stick with the topic established by the initial post.

**Comment by archo5** [comment]
> If you're curious about how the spec arrived at its present state, there's some history here: http://kvark.github.io/spirv/2021/05/01/spirv-horrors.html But this issue isn't the place to revisit decisions made long ago, after extensive discussion.

Perhaps that link would be more relevant if the post offered an explanation for attempting to introduce alien syntax in a text-based shading language. 🤷

I've seen the post before though and don't mind agreeing with the thought that SPIR-V can be messy or unsuited for being used as a cross-language IR in many ways.

At the end of the day though I'm just hoping for one thing:

- a text-based language that can be targeted by covering most of the differences between at least various WebGL GLSL versions with custom macros and basic text editing tools (built-in or custom basic preprocessor/String.replaceAll/a very short JS library etc.)

A way to allow to do initial parsing and optimizations (on IR?) offline (bypassing them at runtime) would be good too if possible, but that would indeed be a different topic.

**Comment by kvark** [comment]
I'm sorry, WGSL can't be targeted by simply preprocessing GLSL. This train has left the station years ago. WGSL is a separate language with it's own design constraints. It's not a SPIR-V text form either.

If you need to target both WebGL and WebGPU (presumably, because you are using something like storage buffers or compute), then you can pick the primary target and generate the shaders for the secondary one. For example, by picking WebGPU as the primary target I can write WGSL shaders, and then convert WGSL -> GLSL ES 3.0 for running WebGL. And the opposite - by keeping your WebGL code path as primary you can generate GLSL -> WGSL shaders at run-time.

Both conversion paths are supported by [naga](https://github.com/gfx-rs/naga) (with caveats), which is trivial to carry as a WASM module. The WGSL->GLSL path is also supported by [Tint](https://dawn.googlesource.com/tint/).

**Comment by archo5** [comment]
> This train has left the station years ago. WGSL is a separate language with it's own design constraints.

AFAICT the rationale for this has not been explained anywhere. I've looked in the following places (and going 1-2 levels deep using the links found there):

* https://gpuweb.github.io/gpuweb/wgsl/#goals
* https://github.com/gpuweb/gpuweb/issues/610#issuecomment-699753812
  * by the way, `that's a fairly big document` (referring to "WebGPU GLSL flavor") sounds strange when [this](https://gpuweb.github.io/gpuweb/wgsl/) doc exists
* https://github.com/gpuweb/gpuweb/pull/562/files
* some random reddit/HN comments

> presumably, because you are using something like storage buffers or compute

The more likely use case in the first few years seems to be primarily targeting a WebGL1/2 feature set and trying to support WebGPU (+/- same feature set) for the performance benefits or minor graphics upgrades.
The occasional compute shader may exist (for e.g. skinning meshes or post-processing) but I wouldn't say that's likely to be the bulk of the code.

Once the majority of browsers and devices support WebGPU by default (as in, no need to turn on some "experimental feature" flag), we might see a 2-3 year transition period into WebGPU-majority then. WebGL2 has become truly cross-platform only just recently and engine/game feature sets reflect this (despite supporting WebGL2 here and there, the majority of the workload was WebGL1-compatible).

> which is trivial to carry as a WASM module

Are there any plans to create official distributions for one-way conversion libraries (or just GLSL -> WGSL)? So far I've only found the [wasm-naga](https://github.com/pjoe/wasm-naga) ([npm info](https://libraries.io/npm/wasm-naga)) project, which doesn't seem to even contain this option and isn't kept up to date either.

I wouldn't call maintaining my own fork of naga trivial, though. There's enough things that require this already, and as some would say, there's "no appetite" for adding more such libraries to the project. 😏

**Comment by dneto0** [comment]
> Are there any plans to create official distributions for one-way conversion libraries (or just GLSL -> WGSL)?

What does "official" mean to you?  It's unlikely W3C would take on that work.  Like @kvark said, they could be outputs of Naga or Tint projects, and that is probably as close as you'll get.   I filed corresopnding Tint issue https://crbug.com/tint/1326

**Comment by archo5** [comment]
> I filed corresopnding Tint issue https://crbug.com/tint/1326

Thanks!

> What does "official" mean to you?

Basically something that's relatively dependable:

* updates being published to the conversion library whenever Naga/Tint receives bugfixes (not necessarily immediately but within some reasonable timeframe);
* all library releases undergoing at least some basic testing;
* Naga/Tint development taking the needs of these libraries into account;
* some way to report issues and somebody responding to these reports.

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2022-05-17-minutes#-initial-internal-feedback-2213-">WGSL meeting minutes 2022-05-17</a></summary>



* Google: Thinks all items have been addressed.
* KG: Is it satisfied, probably but need to ask Apple.


</details>

**Comment by litherum** [comment]
All these items have been moved to other issues (and closed).

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2022-05-24-minutes#-initial-internal-feedback-2213-">WGSL meeting minutes 2022-05-24</a></summary>



* Previously:
    * Google: Thinks all items have been addressed.
    * KG: Is it satisfied, probably but need to ask Apple.
* MM: **Closed**


</details>
